test/abstract_syntax_tree_location_data.coffee

```diff
@@ -4456,3 +4456,384 @@ test "AST as expected for If node", ->
       end:
         line: 5
         column: 1
+
+test "AST as expected for While node", ->
+  testAstLocationData 'loop 1',
+    type: 'WhileStatement'
+    test:
+      start: 0
```

Comment:
So I'd consider the inclusion of the generated `BooleanLiteral` (`true`) in the AST for `loop` to be worth questioning:

In general, we don't want to include any AST nodes that aren't actually present in the original source since we can't output AST that violates location data nesting/ordering (its location data needs to look "real"/sequential or else both Prettier and ESLint will get upset)

In this case, we can get away with assigning the location data of the `loop` keyword (so here, `range: [0, 4]`) to the generated `BooleanLiteral` and that still results in an AST with location data that makes sense

And the upside of including the generated `true` condition in the AST is that it's more compatible - it both avoids potential incompatibilities due to having a `null` `test` (which would never happen in a real Babel `while` AST) and lets things that might be interested in inspecting the `while`'s condition (eg ESLint code path analysis) know that it's effectively a constant `true` condition

So it's a question of weighing the upsides against the general downsides of including any generated nodes in the AST. In this case, I think it's worth including since it makes the generated AST for `loop` statements pretty much compatible with their corresponding Babel `while` ASTs